### ISSUE TITLE
JSONPath parser doesn't handle checks for nested properties

### ISSUE THREAD

**Issue by tillig** [stoplightio/spectral#1287]
**Describe the bug**

I'm trying to create a rule that requires all parameters to have an `example` or `examples` field _if_ it's not a `$ref` to some predefined entity.

The rule looks like this:

```yaml
  required-examples-parameters:
    description: Parameters should include examples unless they reference a schema that might already have examples.
    given: $..[?(@parentProperty !== 'links' && @.parameters)]['parameters'].[?(@.in && !@.schema.$ref)]
    then:
      function: xor
      functionOptions:
        properties:
          - example
          - examples
```

The rule does not properly handle the check for `!@.schema.$ref` via JSONPath though it seems to be a valid query.

**To Reproduce**

Create an OpenAPI `testdoc.yml` doc like this:

```yaml
openapi: '3.0.2'
info:
  title: Valid API
  version: '1.0'
paths:
  /root/v1/test/{id}:
    get:
      operationId: test-route-by-id
      description: Gets a thing by ID.
      parameters:
        - in: path
          name: id
          description: An ID for the fake test object.
          schema:
            $ref: '#/components/schemas/Identity'
          required: true
      responses:
        '200':
          description: OK
      tags:
        - test
components:
  schemas:
    Identity:
      type: integer
      example: 100
```

Now create a `.spectral.yml` like this:

```yaml
extends: spectral:oas
formats: ["oas3"]
rules:
  required-examples-parameters:
    description: Parameters should include examples unless they reference a schema that might already have examples.
    given: $..[?(@parentProperty !== 'links' && @.parameters)]['parameters'].[?(@.in && !@.schema.$ref)]
    then:
      function: xor
      functionOptions:
        properties:
          - example
          - examples
```

Put both of those documents in the same folder and...

`spectral lint testdoc.yml`

You will see a violation of the `required-examples-parameters` rule. **That rule shouldn't be violated** because the error is happening on a path parameter that has a `$ref` in the schema. It shouldn't match the JSONPath but it does.

**Expected behavior**

No violations of the `required-examples-parameters` rule for the test document.

**Environment**
 - Spectral CLI 5.4.0
 - OS: MacOS

**Additional context**

If you take the JSON version of that doc...

```json
{
  "openapi": "3.0.2",
  "info": {
    "title": "Valid API",
    "version": "1.0"
  },
  "paths": {
    "/root/v1/test/{id}": {
      "get": {
        "operationId": "test-route-by-id",
        "description": "Gets a thing by ID.",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "description": "An ID for the fake test object.",
            "schema": {
              "$ref": "#/components/schemas/Identity"
            },
            "required": true
          }
        ],
        "responses": {
          "200": {
            "description": "OK"
          }
        },
        "tags": [
          "test"
        ]
      }
    }
  },
  "components": {
    "schemas": {
      "Identity": {
        "type": "integer",
        "example": 100
      }
    }
  }
}
```

And go pop that into https://jsonpath.com, try the expression:

`$..[?(@parentProperty !== 'links' && @.parameters)]['parameters'].[?(@.in && !@.schema.$ref)]`

It doesn't find anything. If you remove the `!` so it's:

`$..[?(@parentProperty !== 'links' && @.parameters)]['parameters'].[?(@.in && @.schema.$ref)]`

It finds the one property there.

I gather that what I'm trying to do is valid JSONPath, but it's not something the Spectral JSONPath evaluator is handling right.

Is there an alternative expression I could use to select parameters that aren't referencing other schema?

**Comment by nulltoken** [comment]
@tillig Spectral operates on dereferenced documents. You might be willing to decorate your custom rule with `resolved: false`
(cf. https://github.com/stoplightio/spectral/issues/436 and https://meta.stoplight.io/docs/spectral/docs/guides/4-custom-rulesets.md)

**Comment by tillig** [comment]
Oh my goodness, that's the ticket! I need to RTFM better, somehow I missed that. Thank you for the help and your patience, I very much appreciate it.

**Comment by nulltoken** [comment]
>Oh my goodness, that's the ticket! I need to RTFM better

@tillig Or maybe we need to WTFM better 😉 More seriously, I happen to know about this option, but it's quite an obscure/advanced one.

If you happen to think about a better wording or even a better place in the documentation, we'd be very eager to make the doco more usable.

**Comment by tillig** [comment]
Cool beans. If I think of some way to improve, I'll submit a PR. I did notice a couple of places where some links from one doc to another could help, so I'll PR those.
